File: server/src/retrieval/complexity.js
// Cheap heuristic (no LLM call) classifying question complexity into the
// source-count tiers the product spec requires: simple/medium/complex/research.
// Mirrors Chat.jsx's client-side isComplexPrompt but produces a tier instead
// of a binary fast/deep routing decision — used purely to size retrieval,
// never to decide which pipeline runs.
const RESEARCH_KEYWORDS = /\b(research|in depth|comprehensive|thorough|literature|survey of|state of the art)\b/i;
const COMPLEX_KEYWORDS = /\b(compare|comparison|vs\.?|versus|difference between|pros and cons|analyze|analyse|evaluate|trade-?offs?)\b/i;

const TIERS = {
  simple: { minSources: 3, topK: 3 },
  medium: { minSources: 5, topK: 5 },
  complex: { minSources: 8, topK: 15 },
  research: { minSources: 10, topK: 20 },
};

export function classifyComplexity(query) {
  const trimmed = (query || '').trim();
  const wordCount = trimmed.split(/\s+/).filter(Boolean).length;
  const questionMarks = (trimmed.match(/\?/g) || []).length;

  if (RESEARCH_KEYWORDS.test(trimmed) || wordCount > 60) return 'research';
  if (COMPLEX_KEYWORDS.test(trimmed) || wordCount > 25 || questionMarks > 1) return 'complex';
  if (wordCount > 10) return 'medium';
  return 'simple';
}

export function tierFor(query) {
  return TIERS[classifyComplexity(query)];
}

File: server/app/retrieval/complexity.py
import re

RESEARCH_KEYWORDS = re.compile(
    r"\b(research|in depth|comprehensive|thorough|literature|survey of|state of the art)\b", re.I
)
COMPLEX_KEYWORDS = re.compile(
    r"\b(compare|comparison|vs\.?|versus|difference between|pros and cons|analyze|analyse|evaluate|trade-?offs?)\b",
    re.I,
)

TIERS = {
    "simple": {"minSources": 3, "topK": 5},
    "medium": {"minSources": 5, "topK": 8},
    "complex": {"minSources": 8, "topK": 12},
    "research": {"minSources": 10, "topK": 15},
}


def classify_complexity(query: str) -> str:
    trimmed = (query or "").strip()
    word_count = len([w for w in trimmed.split() if w])
    question_marks = trimmed.count("?")

    if RESEARCH_KEYWORDS.search(trimmed) or word_count > 60:
        return "research"
    if COMPLEX_KEYWORDS.search(trimmed) or word_count > 25 or question_marks > 1:
        return "complex"
    if word_count > 10:
        return "medium"
    return "simple"


def tier_for(query: str) -> dict:
    return TIERS[classify_complexity(query)]
